### PR TITLE
Gnosis Safe sign message integration

### PR DESCRIPTION
**Pull request by kaitoo1**
Enable Gnosis Safe users to connect to Gallery by signing a message. 

Changes:
- Added Gnosis Safe as an option to Wallet Selector for signing in and adding a wallet to a user.
- If connecting with Gnosis Safe, front end will prompt the user to submit a tx to sign the message, and listen on-chain for the tx to be approved and executed. Then it will call the backend to validate the signature in order to complete the auth action.
- refactored the pending state screens in the Wallet Selector so that we have more control over the biz logic and UI changes for different wallet types. 
- Also refactored the `initializeAuthPipeline` and `initializeAddWalletPipeline` functions and broke them apart so that we can call different pieces of it individually. This was necessary for Gnosis, where we want to do additional things in the middle of the pipeline like updating the UI to inform the user about next steps.
- created a Wallet Utils file that contains useful functions for wallet specific actions like signing a message, detecting the type of ethirium account, etc

### CODE REVIEW

File: src/components/WalletSelector/authRequestUtils.ts [hunk 1]
```diff
@@ -131,17 +136,57 @@ function isRpcSignatureError(error: Record<string, any>) {
...
+      
       // This keeps the nonce message intact instead of encrypting it for WalletConnect users
       // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
-      return await connector.walletConnectProvider.connector.signPersonalMessage([nonce, address]) as Signature;
+      const signature = await connector.walletConnectProvider.connector.signPersonalMessage([nonce, address]) as Signature;
+      // for gnosis, signature will always be 0x. see Notion for explanation
+
+      // todo create separate signMessage func for gnosis/smart contracts.
+
+      if (library){
```
Comment (Robinnnnn) [inline src/components/WalletSelector/authRequestUtils.ts:158]: random q: do you have prettier turned on?

File: src/components/WalletSelector/authRequestUtils.ts [hunk 2]
```diff
@@ -131,17 +136,57 @@ function isRpcSignatureError(error: Record<string, any>) {
...
+        // create listener that will listen for the SignMsg event on the Gnosis contract
+        const listenForGnosis = new Promise((resolve, reject) => {          
+          gnosisContract.on('SignMsg', async (msgHash:any, event: any, error: any) =>  {
+            console.log('msgHash', msgHash)
+            console.log('event', event)
+            console.log('error', error)
+            
+            if (error) {
+              reject(error)
+            }
```
Comment (Robinnnnn) [inline src/components/WalletSelector/authRequestUtils.ts:174]: i believe we want to `return` after the reject

File: src/components/WalletSelector/authRequestUtils.ts (hunk 1 above)
Comment (kaitoo1) [reply src/components/WalletSelector/authRequestUtils.ts]: turned it on now, should be resolved!

File: src/utils/wallet.ts [hunk 3]
```diff
@@ -1,10 +1,14 @@
...
+  METAMASK: 'MetaMask',
+  WALLETCONNECT: 'WalletConnect',
+  WALLETLINK: 'Coinbase Wallet',
+  GNOSIS_SAFE: 'Gnosis Safe',
 };
 
 // Convert wallet provider name to user friendly name
-export function convertWalletName(name: string) {
+export function getUserFriendlyWalletName(name: string) {
+  console.log(name);
```
Comment (kaitoo1) [inline src/utils/wallet.ts:11]: delete

File: src/components/WalletSelector/walletUtils.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    gnosisSafeContract.on(GNOSIS_SAFE_SIGN_MESSAGE_EVENT_NAME, async (msgHash: any) => {
+      // Upon detecing the SignMsg event, validate that the contract signed the message
+      const magicValue = await gnosisSafeContract.isValidSignature(messageHash, '0x');
+      const messageWasSigned = magicValue === GNOSIS_VALID_SIGNATURE_MAGIC_VALUE;
+
+      if (messageWasSigned) {
+        resolve(msgHash);
+      }
+
+      console.log('messageWasSigned', messageWasSigned);
```
Comment (kaitoo1) [inline src/components/WalletSelector/walletUtils.ts:132]: delete

**Comment by Robinnnnn** [comment]
looks like there's a vercel build error
image

File: src/components/WalletSelector/AddWalletPending/AddWalletPending.tsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+function AddWalletPending({
+  pendingWallet,
+  userFriendlyWalletName,
+  setDetectedError,
+  walletName,
+}: Props) {
+  if (walletName === GNOSIS_SAFE) {
+    return (
+      <AddWalletPendingGnosisSafe
+        {...{ pendingWallet, userFriendlyWalletName, setDetectedError }}
```
Comment (Robinnnnn) [inline src/components/WalletSelector/AddWalletPending/AddWalletPending.tsx:25]: hmm this unnecessarily creates a new object. could we just keep it explicit with `pendingWallet={pendingWallet}` etc?

File: src/components/WalletSelector/AddWalletPending/AddWalletPending.tsx [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+      />
+    );
+  }
+
+  return (
+    <AddWalletPendingDefault
+      {...{ pendingWallet, userFriendlyWalletName, setDetectedError, walletName }}
```
Comment (Robinnnnn) [inline src/components/WalletSelector/AddWalletPending/AddWalletPending.tsx:32]: same comment

File: src/components/WalletSelector/AddWalletPending/AddWalletPendingGnosisSafe.tsx [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        <BodyRegular color={colors.gray50}>
+          Follow the prompts in the Gnosis app to sign the message.
+        </BodyRegular>
+        <Spacer height={8} />
+        <BodyRegular color={colors.gray50}>Do not close this window.</BodyRegular>
+      </div>
+    );
+  }
+
+  if (pendingState === LISTENING_ONCHAIN) {
```
Comment (Robinnnnn) [inline src/components/WalletSelector/AddWalletPending/AddWalletPendingGnosisSafe.tsx:157]: love these states

File: src/components/WalletSelector/AddWalletPending/AddWalletPendingGnosisSafe.tsx [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+    return (
+      <div>
+        <TitleMedium>Connect with {userFriendlyWalletName}</TitleMedium>
+        <Spacer height={24} />
+        <BodyRegular color={colors.gray50}>
+          Connecting with Gnosis requires an on chain transaction.
+        </BodyRegular>
+        <Spacer height={8} />
+        <BodyRegular color={colors.gray50}>
+          Transaction submitted - this must now be confirmed and executed by the Gnosis Safe owners.
```
Comment (Robinnnnn) [inline src/components/WalletSelector/AddWalletPending/AddWalletPendingGnosisSafe.tsx:167]: copy update

File: src/components/WalletSelector/AuthenticateWalletPending/AuthenticateWalletPending.tsx [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+function AuthenticateWalletPending({
+  pendingWallet,
+  userFriendlyWalletName,
+  setDetectedError,
+  walletName,
+}: Props) {
+  if (walletName === GNOSIS_SAFE) {
+    return (
+      <AuthenticateWalletPendingGnosisSafe
+        {...{ pendingWallet, userFriendlyWalletName, setDetectedError }}
```
Comment (Robinnnnn) [inline src/components/WalletSelector/AuthenticateWalletPending/AuthenticateWalletPending.tsx:31]: same comment

File: src/components/WalletSelector/AuthenticateWalletPending/AuthenticateWalletPendingGnosisSafe.tsx [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        <BodyRegular color={colors.gray50}>
+          Follow the prompts in the Gnosis Safe app to submit a transaction to sign the message.
+        </BodyRegular>
+        <Spacer height={32} />
+        <BodyRegular color={colors.gray50}>Do not close this window.</BodyRegular>
+      </div>
+    );
+  }
+
+  if (pendingState === LISTENING_ONCHAIN) {
```
Comment (Robinnnnn) [inline src/components/WalletSelector/AuthenticateWalletPending/AuthenticateWalletPendingGnosisSafe.tsx:109]: this logic seems to be shared with the other file – `AddWalletPendingGnosisSafe.tsx`. is it worth consolidating?

File: src/components/WalletSelector/walletUtils.ts [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+  address: string,
+  nonce: string,
+  connector: AbstractConnector,
+  signer: JsonRpcSigner,
+  library?: Web3Provider
+): Promise<Signature> {
+  const accountType = getEthereumAccountType(connector);
+
+  if (accountType === CONTRACT_ACCOUNT) {
+    // if theres a nonce in local storage, prompt to try again
```
Comment (Robinnnnn) [inline src/components/WalletSelector/walletUtils.ts:44]: is this a TODO, or is that what's actually happening here?

Comment (kaitoo1) [reply src/components/WalletSelector/walletUtils.ts:44]: good call, this is a todo for optimization

File: src/components/WalletSelector/AddWalletPending/AddWalletPending.tsx (hunk 5 above)
Comment (kaitoo1) [reply src/components/WalletSelector/AddWalletPending/AddWalletPending.tsx]: done 
dang I thought I was being clever lol

File: src/components/WalletSelector/AddWalletPending/AddWalletPendingGnosisSafe.tsx (hunk 8 above)
Comment (kaitoo1) [reply src/components/WalletSelector/AddWalletPending/AddWalletPendingGnosisSafe.tsx]: done

File: src/components/WalletSelector/AuthenticateWalletPending/AuthenticateWalletPending.tsx (hunk 9 above)
Comment (kaitoo1) [reply src/components/WalletSelector/AuthenticateWalletPending/AuthenticateWalletPending.tsx]: done

File: src/components/WalletSelector/AuthenticateWalletPending/AuthenticateWalletPendingGnosisSafe.tsx (hunk 10 above)
Comment (kaitoo1) [reply src/components/WalletSelector/AuthenticateWalletPending/AuthenticateWalletPendingGnosisSafe.tsx]: moved shared code into a new component `GnosisSafePendingMessage` that contains Gnosis Safe specific screens for the pending state, used for both Auth and Add Wallet flows

**Pull request merged**